sandbox/apparmor/apparmor.go

```diff
@@ -48,6 +49,401 @@ func ValidateNoAppArmorRegexp(s string) error {
+					continue
+				}
+			}
+			// snippet up to this character so far - again since this is the
+			// common prefix we can just use the first pattern
+			snippet := string(excludePatterns[0][:i])
+			res := fmt.Sprintf("%s%s[^%c]**%s\n", opts.Prefix, snippet, c, opts.Suffix)
+
+			builder.WriteString(res)
+
+			// save this as the most recent common sub string pattern
+			lastCommonPrefix = snippet + string(c)
+		case i == endCommonPrefix:
+			// this is the last bit of the common substring, so get all the
+			// characters for each pattern and just make a negative match for
+			// all of those characters
+
+			// note we use a map because there could be duplicate characters,
+			// for example we could have 3 strings like "abcd1", "abcd2", "abceee",
+			// where after the first common characters among the 3, there is
+			// another common character with first 2 exclude patterns not in the
+			// third
+			chars := map[rune]bool{}
+			for _, pattern := range excludePatterns {
+				if i >= len(pattern) {
+					continue
+				}
+				chars[pattern[i]] = true
+			}
+			if len(chars) == 0 {
+				// shouldn't happen in practice, we check for duplicates above
+				// but just be on the safe side
+				return "", fmt.Errorf("internal erorr: all excluded patterns are the same")
+			}
+
+			negGroup := []rune("[^")
+			for c := range chars {
+				negGroup = append(negGroup, c)
+			}
+			// make sure the runs are sorted for consistency
+			sort.Slice(negGroup, func(i, j int) bool {
+				return negGroup[i] < negGroup[j]
+			})
+			negGroup = append(negGroup, ']')
+
+			res := fmt.Sprintf("%s%s%s**%s\n", opts.Prefix, lastCommonPrefix, string(negGroup), opts.Suffix)
+			builder.WriteString(res)
+
+		case i > endCommonPrefix:
+			// we break and use a separate loop because here we now need to
+			// handle indexes per pattern because "*" could be in different
+			// locations in each of the exclude patterns and we can no longer
+			// use one index for all patterns
+			break charLoop
+		}
+	}
+
+	// each pattern starts at i
+	indexPerPattern := make([]int, len(excludePatterns))
+	for i2 := range excludePatterns {
+		indexPerPattern[i2] = i
+	}
+
+	// this loop iterates up until the longestStrLen is reached or until we
+	// reach only one pattern left which then collapses into a simpler special
+	// case
+	for j := i; j < longestStrLen; j++ {
+		// check what patterns still have characters left
+		stillPresentPatterns := map[int][]rune{}
+		for patternIndex, excludePattern := range excludePatterns {
+			charIndex := indexPerPattern[patternIndex]
+			if charIndex >= len(excludePattern) {
+				continue
+			}
+			stillPresentPatterns[patternIndex] = excludePattern
+		}
+
+		// maybe we only have one pattern left which means we can take a
+		// shortcut
+		if len(stillPresentPatterns) == 1 {
+			// there is only one group left from here and we can do the simple
+			// generation with this pattern for the rest of the rules
```

Comment:
I think we should be able to call `GenerateAAREExclusionPatternsSingle` here, if we add the common exclude pattern computed so far to the opts.Prefix (of course, we need to pass a copy of the `opts` variable, not to modify our own one)

Reply:
hmm so the issue with trying to implement this is that the loop for generateAAREExclusionPatternsSingle starts at 1 to skip the "/" in the normal exclude pattern case, however to re-use that here, we would need to start at 0 instead. I considered adding an option to not skip the first character / allow non-absolute filepaths but doing so would require more tests so I think I will just leave a TODO about this optimization instead